<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<title>Backtransform morphospace</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" >
<link rel="shortcut icon" href="../../img/favicon.png"/>
<link href="../../css/stylesheet2.css" rel="stylesheet" type="text/css">

<body>
	<div id='container' class='container'>

		<div id='header' class='header' >
			<div style='position:relative;' >
				<img id='header_image' style='float:right;' src='../../img/headers/Canada Goose.jpg' />
				<div id='header_image_caption' style='color:white' class='header_image_caption' >
					Canada Goose
				</div>
			</div>
		</div>

		<div id='linksidebar' class='linksidebar' >
			<div class='linksidebar_mainlink' >
				<a class='linksidebar_link' href='../../about_me.html'>About me</a>
			</div>
			<div class='linksidebar_mainlink' >
				<a class='linksidebar_link' href='../../current_projects.html'>Current projects</a>
			</div>
			<div class='linksidebar_mainlink' >
				<a class='linksidebar_link' href='../../software.html'>Software</a>
			</div>
			<div class='linksidebar_mainlink' >
				<a class='linksidebar_link' href='../../tutorials/visualization3d.html'>Tutorials</a>
			</div>
			<div class='linksidebar_sublink' >
				<a class='linksidebar_link' href='../../tutorials/visualization3d.html'>3D Visualization</a>
			</div>
			<div class='linksidebar_sublink' >
				<a class='linksidebar_link' href='../../tutorials/morphometrics.html'>Shape analysis</a>
			</div>
			<div class='linksidebar_sub2link_u' >
				<a class='linksidebar_link_u' href='../../tutorials/morphometrics/backtransform.html'>Create a backtransform morphospace</a>
			</div>
			<div class='linksidebar_sublink' >
				<a class='linksidebar_link' href='../../tutorials/motion.html'>Motion analysis</a>
			</div>
			<div class='linksidebar_sublink' >
				<a class='linksidebar_link' href='../../tutorials/mechanisms.html'>Mechanism models</a>
			</div>
			<div class='linksidebar_sublink' >
				<a class='linksidebar_link' href='../../tutorials/stereomorph.html'>StereoMorph User Guide</a>
			</div>
			<div class='linksidebar_sublink' >
				<a class='linksidebar_link' href='../../tutorials/knex/large_u_joint.html'>K'nex models</a>
			</div>
			<div class='linksidebar_mainlink' >
				<a class='linksidebar_link' href='../../contact.html'>Contact</a>
			</div>
			<div class='linksidebar_mainlink' >
				<a class='linksidebar_link' href='../../cv.html'>CV</a>
			</div>
		</div>

		<div id='maincontent' class='maincontent' >

			<div id='pagetrail' class='pagetrail' ><a href="../../about_me.html">home</a>&nbsp;&nbsp;&nbsp;>&nbsp;&nbsp;&nbsp;<a href="../../tutorials/visualization3d.html">tutorials</a>&nbsp;&nbsp;&nbsp;>&nbsp;&nbsp;&nbsp;<a href="../../tutorials/morphometrics.html">shape analysis</a>&nbsp;&nbsp;&nbsp;>&nbsp;&nbsp;&nbsp;create a backtransform morphospace</div>


<div class='maincontentfill'>
<h1>Create a backtransform morphospace</h1>

<p>
This tutorial will show you how to create a "backtransform morphospace" 
(both <a href="#background_morphospace" >non-phylogenetic</a> and <a href="#background_morphospace_ppca">phylogenetic</a>) 
in <a href="http://rweb.quant.ku.edu/cran/" target="_blank" >R</a> 
(below). A backtransform morphospace takes a standard morphospace plot, which 
groups objects or biological specimens according to how similar they are in shape, 
and adds representative shapes in a grid pattern across the plot to show intuitively 
how shape varies across morphospace (<a href='http://onlinelibrary.wiley.com/doi/10.1111/1365-2435.12890/full' target='_blank' >Olsen 2017</a>).
</p>

<div class='p_image_middle_div' style="width:550px;">
	<div><a href='../../img/tutorials/morphometrics/backtransform_morphospace.jpg' target='_blank'>
		<img width='550px' src='../../img/tutorials/morphometrics/backtransform_morphospace.jpg' /></a></div>
	<div class='p_image_caption' style='text-align:center;' >
		Backtransform morphospace of waterfowl beak shapes
	</div>
</div>

<p>
These shapes placed into the background of the plot are "backtransformations" from a principal 
coordinates analysis, which identifies the major axes of variation with a shape dataset. 
Importantly, these backtranform shapes are not actual shapes within the input dataset. 
Rather they are theoretical shapes that represent a combination of two particular PC (principal component) scores 
and the mean of all the other PC scores. An alternative way to think of these 
backtransform shapes is that they are the shape an object or specimen would have 
if it were at a particular point in a 2D morphospace (e.g. at a particular 
value of PC1 and PC2) and at the average position along all other PC axes.
</p>

<p>
If most of your shape variation can be condensed 
into a few PC scores then these theoretical shapes will correspond well to the 
shapes in the input dataset. A useful feature of a backtransform morphospace is 
that it shows what shapes look like throughout the entire morphospace, rather 
than just where you have samples, to create an even distribution for visualization.
</p>

<p>
If you are using this code for a publication please cite the following paper:
</p>

<div style='margin:0 0 20px 20px;padding-left:10px;'>
Olsen AM (2017). Feeding ecology is the primary driver of beak shape diversification in waterfowl. 
<em>Functional Ecology</em>. 31(10):1985-1995. DOI: <a href='https://doi.org/10.1111/1365-2435.12890' target='_blank'>10.1111/1365-2435.12890</a>. Get <a href='http://onlinelibrary.wiley.com/doi/10.1111/1365-2435.12890/full' target='_blank' >via Wiley</a>, 
<a href='https://www.researchgate.net/publication/317316889_Feeding_ecology_is_the_primary_driver_of_beak_shape_diversification_in_waterfowl' target='_blank' > via ResearchGate</a>.
</div>

<h2>Preliminary steps</h2>

<p>
All of the code in this tutorial are R commands. 
So be sure you have <a href="http://rweb.quant.ku.edu/cran/" target="_blank" >R</a>
installed on your system. If you do not have R installed you can find installation 
instructions <a href="https://cran.r-project.org" target="_blank" >here</a>.
</p>

<p>
You will also need to have the R packages <a href="https://CRAN.R-project.org/package=geomorph" target="_blank" >geomorph</a> 
and <a href="https://CRAN.R-project.org/package=StereoMorph" target="_blank" >StereoMorph</a>
installed. 
The geomorph package will allow you to perform the Generalized Procrustes Analysis and the 
StereoMorph package has the function for creating the backtransform morphospace. 
For StereoMorph <em>be sure that you have a version equal to or greater than 1.6.2 installed</em> since this has the btShapes function. 
If you have not installed 
these packages you can install them using the <span class='in_line_code'>install.packages</span> function.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Install R packages (if not installed already)</span>
<span class="code_function" >install.packages</span>(<span class="code_quote" >'geomorph'</span>, dependencies=TRUE)
<span class="code_function" >install.packages</span>(<span class="code_quote" >'StereoMorph'</span>, dependencies=TRUE)
</pre></div>

<p>
Once the package is installed, load it into the current R environment using the 
<span class='in_line_code'>library</span> function.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Load the packages</span>
<span class="code_function" >library</span>(geomorph)
<span class="code_function" >library</span>(StereoMorph)
</pre></div>

<a name="background_morphospace" ></a>
<h2>Creating a (non-phylogenetic) backtransform morphospace using 3D shape data</h2>

<p>
The main part of this tutorial will first demonstrate making a backtransform morphospace 
using 3D shape data and a non-phylogenetic principal components analysis (PCA). 
Creating a <a href="#background_morphospace_ppca">phylogenetic morphospace</a> or using 
<a href="#background_morphospace_2d">2D shape data</a> requires only a few modifications 
to this code, which are detailed after this section.
</p>

<p>
First we need shape data to plot. These should be landmarks or curve points 
(either two- or three-dimensional) saved in a three-dimensional array in which 
the first array dimension is the point names, the second dimension is the xy-coordinates 
(or xyz-coordinates for 3D data), and the third dimension is each individual/species/etc. 
For this tutorial we'll use 3D landmark and curve points collected 
from waterfowl bird beaks, originally published in  <a href='http://onlinelibrary.wiley.com/doi/10.1111/1365-2435.12890/full' target='_blank' >this paper</a> 
looking at the correlation between beak shape and diet in waterfowl. You can 
download these data from <a href='http://datadryad.org/resource/doi:10.5061/dryad.42s0c' target='_blank' >datadryad</a> by clicking <a href='http://datadryad.org/bitstream/handle/10255/dryad.143404/Beak%20shape%20by%20species.zip?sequence=1' target='_blank'>here</a>. 
Unzip this folder and move it to your current working directory.
</p>

<p>
Read the example shape data from the 'Beak shape by species' folder.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Read all files in "Beak shape by species" folder into list</span>
read_files <- <span class="code_function" >lapply</span>(<span class="code_function" >paste0</span>(<span class="code_quote" >'Beak shape by species/'</span>, 
	<span class="code_function" >list.files</span>(<span class="code_quote" >'Beak shape by species'</span>)), <span class="code_quote" >'read.table'</span>)
</pre></div>

<p>
Convert the shape data from list of matrices into a three-dimensional array, 
specifying names along the first and third dimension of the array. This should 
create a 150 x 3 x 51 dimension array.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Convert list of shape data into array</span>
lm_array <- <span class="code_function" >array</span>(<span class="code_function" >unlist</span>(read_files), dim=<span class="code_function" >c</span>(<span class="code_function" >dim</span>(read_files[[<span class="code_numeric" >1</span>]]), 
	<span class="code_function" >length</span>(read_files)), dimnames=<span class="code_function" >list</span>(<span class="code_function" >rownames</span>(read_files[[<span class="code_numeric" >1</span>]]), 
	NULL, <span class="code_function" >gsub</span>(<span class="code_quote" >'[.]txt$'</span>, <span class="code_quote" >''</span>, <span class="code_function" >list.files</span>(<span class="code_quote" >'Beak shape by species'</span>))))
</pre></div>

<p>
Perform <a href="https://en.wikipedia.org/wiki/Procrustes_analysis" target="_blank" >Generalized Procrustes Analysis</a> 
(GPA) using the <a href="https://CRAN.R-project.org/package=geomorph" target="_blank" >geomorph</a> function 
<a href="https://www.rdocumentation.org/packages/geomorph/versions/3.0.3/topics/gpagen" target="_blank" >gpagen</a> 
to obtain Procrustes coordinates.
GPA scales each shape to the same centroid size and 
optimally aligns the shapes (using translation and rotation) based on homologous 
points to remove differences due simply to differences in position and orientation.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Get generalized Procrustes coordinates</span>
gpa_array <- <span class="code_function" >gpagen</span>(lm_array)$coords
</pre></div>

<p>
Perform a (non-phylogenetic) principal components analysis (PCA) to identify the 
major axes of shape variation.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Convert array to matrix for PCA</span>
gpa_mat <- <span class="code_function" >t</span>(<span class="code_function" >apply</span>(gpa_array, <span class="code_numeric" >3</span>, <span class="code_function" >function</span>(y) <span class="code_function" >matrix</span>(<span class="code_function" >t</span>(y), <span class="code_numeric" >1</span>)))

<span class="code_comment" ># Perform non-phylogenetic PCA</span>
resEig <- <span class="code_function" >eigen</span>(<span class="code_function" >cov</span>(gpa_mat))

<span class="code_comment" ># Get PC scores</span>
scores <- gpa_mat %*% resEig$vectors

<span class="code_comment" ># Get percent variance explained along each axis</span>
per_var <- (resEig$values / <span class="code_function" >sum</span>(resEig$values))*<span class="code_numeric" >100</span>
</pre></div>

<p>
In order to plot each backtransform shape we need to create a function that 
will take our shape coordinates and produce a graphic of 
the specimen or object. For example, the function could connect a series of neighboring landmarks 
to create an outline. How you draw the shape data depends on what kind of shape data you have 
and how you want the graphic to look, so you will have to create this function 
yourself. But the function included in this example should serve as a useful template.
</p>

<p>
You can name the function whatever you'd like. The example below is named <span class='in_line_code'>plot_beak_lateral</span>.
This function takes curve points along the bottom and top margin of the beak and 
creates a silhouette of the beak from a lateral view. The shape data that are input to this function are  
3D so this function just takes a simple projection of the 3D shape onto the midline body plane. 
You could rotate the entire shape before taking the projection to get a 
top-down view instead of a lateral view.
</p>

<p>
The first two parameters to this function have to be the following:
</p>

<ul>
	<li>xy: a 2-column matrix of the PC scores where each backtransform shape will be drawn</li>
	<li>coor: a matrix of the shape coordinates for each backtransform shape (2-column for 2D and 3-column for 3D)</li>
</ul>

<p>
You can include any number of extra parameters that you would like to 
customize the shape drawing. In the example below there is a size parameter to 
adjust the overall size of each backtransform shape and a color parameter to specify the 
fill color of the beaks. The extra parameters that you specify will be passed 
along through the <span class='in_line_code'>btShapes</span> function (the function you create will be called 
by <span class='in_line_code'>btShapes</span>) so they should differ from the input parameters to <span class='in_line_code'>btShapes</span>.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Define function to draw shape</span>
plot_beak_lateral <- <span class="code_function" >function</span>(xy, coor, size=<span class="code_numeric" >1</span>, col=<span class="code_quote" >'black'</span>){

	<span class="code_comment" ># If 3D, rotate points about x-axis using 3D rotation matrix</span>
	if(<span class="code_function" >ncol</span>(coor) == <span class="code_numeric" >3</span>){
		coor <- coor %*% <span class="code_function" >matrix</span>(<span class="code_function" >c</span>(<span class="code_numeric" >1</span>,<span class="code_numeric" >0</span>,<span class="code_numeric" >0</span>, <span class="code_numeric" >0</span>,<span class="code_function" >cos</span>(-<span class="code_numeric" >pi</span>/<span class="code_numeric" >2</span>),<span class="code_function" >sin</span>(-<span class="code_numeric" >pi</span>/<span class="code_numeric" >2</span>), 
			<span class="code_numeric" >0</span>,-<span class="code_function" >sin</span>(-<span class="code_numeric" >pi</span>/<span class="code_numeric" >2</span>),<span class="code_function" >cos</span>(-<span class="code_numeric" >pi</span>/<span class="code_numeric" >2</span>)), nrow=<span class="code_numeric" >3</span>, ncol=<span class="code_numeric" >3</span>)
	}

	<span class="code_comment" ># Get just x,y coordinates (orthographic projection into xy-plane)</span>
	coor <- coor[, <span class="code_numeric" >1</span>:<span class="code_numeric" >2</span>]

	<span class="code_comment" ># Get plot aspect ratio</span>
	w <- <span class="code_function" >par</span>(<span class="code_quote" >'pin'</span>)[<span class="code_numeric" >1</span>]/<span class="code_function" >diff</span>(<span class="code_function" >par</span>(<span class="code_quote" >'usr'</span>)[<span class="code_numeric" >1</span>:<span class="code_numeric" >2</span>])
	h <- <span class="code_function" >par</span>(<span class="code_quote" >'pin'</span>)[<span class="code_numeric" >2</span>]/<span class="code_function" >diff</span>(<span class="code_function" >par</span>(<span class="code_quote" >'usr'</span>)[<span class="code_numeric" >3</span>:<span class="code_numeric" >4</span>])
	asp <- w/h

	<span class="code_comment" ># Correct for plot aspect ratio not necessarily being 1:1</span>
	coor[, <span class="code_numeric" >1</span>] <- coor[, <span class="code_numeric" >1</span>] * (<span class="code_numeric" >1</span>/asp)

	<span class="code_comment" ># Scale points and place back in position</span>
	coor <- coor*size

	<span class="code_comment" ># Center about zero based on range of coordinates</span>
	coor <- coor - <span class="code_function" >matrix</span>(<span class="code_function" >colMeans</span>(<span class="code_function" >apply</span>(coor, <span class="code_numeric" >2</span>, range)), 
		nrow=<span class="code_function" >nrow</span>(coor), ncol=<span class="code_function" >ncol</span>(coor), byrow=TRUE)

	<span class="code_comment" ># Move shape to PC score</span>
	coor <- coor + <span class="code_function" >matrix</span>(xy, <span class="code_function" >nrow</span>(coor), <span class="code_function" >ncol</span>(coor), byrow=TRUE)

	<span class="code_comment" ># Set order in which to draw points to create polygon</span>
	polygon_order <- <span class="code_function" >c</span>(<span class="code_function" >which</span>(<span class="code_function" >grepl</span>(<span class="code_quote" >'upper_bill_culmen'</span>, <span class="code_function" >rownames</span>(coor))),
		<span class="code_function" >rev</span>(<span class="code_function" >which</span>(<span class="code_function" >grepl</span>(<span class="code_quote" >'upper_bill_tomium_L'</span>, <span class="code_function" >rownames</span>(coor)))))

	<span class="code_comment" ># Create filled polygon</span>
	<span class="code_function" >polygon</span>(coor[polygon_order, ], col=col, border=col)
}
</pre></div>

<p>
To use this function in other projects you might want to save this function to its own file,
for example 'plot_beak_lateral.R' and then load the function at the start of your 
code using the <span class='in_line_code'>source</span> function.
</p>

<p>
Set the two PC axes that you want to plot. For this example we'll plot the 
first and second PC axes. These are the two PC axes that form the basis of the grid onto 
which the backtransform shapes are drawn (the shapes will use the average PC scores along all 
the remaining axes).
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Set PCs to plot</span>
pcs <- <span class="code_numeric" >1</span>:<span class="code_numeric" >2</span>
</pre></div>

<p>
For this tutorial we'll create a plot as a PDF file. Open a PDF graphics device 
where the plot will be produced.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Open PDF graphics device</span>
<span class="code_function" >pdf</span>(<span class="code_quote" >'Backtransform PCA.pdf'</span>, width=<span class="code_numeric" >9</span>, height=<span class="code_numeric" >6.5</span>)
</pre></div>

<p>
Create a plot box with axes and axis labels. Use the <span class='in_line_code'>n</span> value for the 
<span class='in_line_code'>type</span> parameter to draw the plot without plotting the scores. That way you can 
add the scores later on top of the backtransform shapes.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Create plot box with axes and axis labels</span>
<span class="code_function" >plot</span>(scores[, pcs], type=<span class="code_quote" >'n'</span>, main=<span class="code_quote" >'Backtransform morphospace'</span>,
	xlab=<span class="code_function" >paste0</span>(<span class="code_quote" >'PC'</span>, pcs[<span class="code_numeric" >1</span>], <span class="code_quote" >' ('</span>, <span class="code_function" >round</span>(per_var[pcs[<span class="code_numeric" >1</span>]]), <span class="code_quote" >'%)'</span>),
	ylab=<span class="code_function" >paste0</span>(<span class="code_quote" >'PC'</span>, pcs[<span class="code_numeric" >2</span>], <span class="code_quote" >' ('</span>, <span class="code_function" >round</span>(per_var[pcs[<span class="code_numeric" >2</span>]]), <span class="code_quote" >'%)'</span>))
</pre></div>

<p>
Plot the backtransform shapes in the plot box using the <span class='in_line_code'>btShapes</span> function.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Plot backtransform shapes</span>
<span class="code_function" >btShapes</span>(scores=scores, vectors=resEig$vectors, fcn=plot_beak_lateral, 
	pcs=pcs, n=<span class="code_function" >c</span>(<span class="code_numeric" >5</span>,<span class="code_numeric" >7</span>), m=<span class="code_function" >dim</span>(lm_array)[<span class="code_numeric" >2</span>], row.names=<span class="code_function" >dimnames</span>(lm_array)[[<span class="code_numeric" >1</span>]], 
	pc.margin=<span class="code_function" >c</span>(<span class="code_numeric" >0.06</span>,<span class="code_numeric" >0.05</span>), size=<span class="code_numeric" >0.5</span>, col=<span class="code_function" >gray</span>(<span class="code_numeric" >0.7</span>))
</pre></div>

<p>
Here are the input parameters to <span class='in_line_code'>btShapes</span>:
</p>

<ul>
	<li><span class='in_line_code'>scores</span>: a matrix of the PCA scores</li>
	<li><span class='in_line_code'>vectors</span>: a matrix of the PCA eigenvectors</li>
	<li><span class='in_line_code'>fcn</span>: the function that will be called by btShapes to draw each shape</li>
	<li><span class='in_line_code'>pcs</span>: the two PC axes for the plot</li>
	<li><span class='in_line_code'>n</span>: a two-element vector giving the number of shapes to draw along the x- and y-axes, respectively</li>
	<li><span class='in_line_code'>m</span>: number of shape dimensions (2 for 2D, 3 for 3D)</li>
	<li><span class='in_line_code'>row.names</span>: a vector giving the names of the landmarks in the same order as the shape array input to the PC analysis</li>
	<li><span class='in_line_code'>pc.margin</span>: a two-element vector giving the margins along the x- and y-axes, 
	respectively, for the range of PC scores that are used to draw the shapes. 
	For example, a value of 0.05 
	will start drawing the backtransform shapes at values 5% greater than the minimum 
	PC score along the corresponding axis and end at 5% less than the maximum PC score. 
	This is useful for making sure the shapes fall within the plot box.</li>
</ul>

<p>
The remaining two input parameters (<span class='in_line_code'>size</span> and <span class='in_line_code'>col</span>) are inputs to the <span class='in_line_code'>plot_beak_lateral</span> 
function so these may differ depending on how you write your shape drawing function.
</p>

<p>
If you'd like, you can then plot the points for each species on top of the backtransform shapes.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Plot points for each species</span>
<span class="code_function" >points</span>(scores[, pcs])
</pre></div>

<p>
Add text labels.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Add text labels</span>
<span class="code_function" >text</span>(scores[, pcs], labels=<span class="code_function" >substr</span>(<span class="code_function" >rownames</span>(scores), <span class="code_numeric" >0</span>, <span class="code_numeric" >3</span>), cex=<span class="code_numeric" >0.8</span>, 
	pos=<span class="code_numeric" >1</span>, offset=<span class="code_numeric" >0.3</span>)
</pre></div>

<p>
Close the PDF graphics device.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Close the PDF graphics device</span>
<span class="code_function" >dev.off</span>()
</pre></div>

<p>
This should produce the following plot:
</p>

<div class='p_image_middle_div' style="width:550px;">
	<div><a href='../../img/tutorials/morphometrics/backtransform_morphospace.jpg' target='_blank'>
		<img width='550px' src='../../img/tutorials/morphometrics/backtransform_morphospace.jpg' /></a></div>
	<div class='p_image_caption' style='text-align:center;' >
		Backtransform morphospace of 3D waterfowl beak shapes
	</div>
</div>



<a name="background_morphospace_ppca" ></a>

<h2>Creating a phylogenetic backtransform morphospace</h2>

<p>
The previous section showed how to use a non-phylogenetic PCA to find the major axes of shape variation. 
However, PCA assumes that each of the data points are independent samples from a population 
(like a standard linear regression). Since sampled taxa are not independent of 
one another but in fact related to each other you might 
want to incorporate the phylogeny for your taxa into the PCA analysis to account for this 
non-independence. Another way to think about this is that if you had a sample in which many 
taxa were all closely related and just a few distantly related taxa, you wouldn't want the many 
closely related specimens to have an outsized influence in determining the PC axes 
(see <a href='https://doi.org/10.1111/j.1558-5646.2009.00804.x' target='_blank' >Revell 2009</a> for a nice explanation of this). 
This section will show you how to create a backtransform morphospace onto a phylogenetic 
PCA (pPCA) plot.
</p>

<p>
Before starting this section make sure you've worked through the previous section. There 
is some overlap in the code needed for the previous section and for this section so it's 
assumed that you have run all of the previous code in the current workspace. 
Before starting also make sure you have the following additional R packages installed.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Install ape and phytools (if not already installed)</span>
<span class="code_function" >install.packages</span>(<span class="code_quote" >'ape'</span>, dependencies=TRUE)
<span class="code_function" >install.packages</span>(<span class="code_quote" >'phytools'</span>, dependencies=TRUE)
</pre></div>

<p>
Then load the packages into the current R environment.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Load the packages</span>
<span class="code_function" >library</span>(ape)
<span class="code_function" >library</span>(phytools)
</pre></div>

<p>
Next, find a tree for your sampled taxa. You can download the tree for this tutorial by 
clicking <a href='https://datadryad.org/bitstream/handle/10255/dryad.143407/Burleigh%2c%20Kimball%2c%20Braun%2c%202015%20ML%20Tree%2c%20pruned.tre?sequence=1' target='_blank' >here</a>. 
Save this file in the web browser using 'File > Save' (be sure the file saves with a <em>'.tre' extension</em> and not '.txt'). 
Then move this tree file ('Burleigh, Kimball, Braun, 2015 ML Tree, pruned.tre') to your 
current R working directory. This tree comes from a larger tree by <a href='https://doi.org/10.1016/j.ympev.2014.12.003' target='_blank' >Burleigh, Kimball & Braun 2015</a> 
and has been pruned down to include just the species in this tutorial.
</p>

<p>
Load the tree using the <span class='in_line_code'>read.tree</span> function from the ape package.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Read tree</span>
tree <- <span class="code_function" >read.tree</span>(file=<span class="code_quote" >'Burleigh, Kimball, Braun, 2015 ML Tree, pruned.tre'</span>)
</pre></div>

<p>
You can use the same <span class='in_line_code'>gpa_mat</span> from the generalized Procrustes coordinates that you created in the 
previous section but this time it will be used to perform a pPCA instead of a PCA, using the 
<span class='in_line_code'>phyl.pca</span> function from the phytools package.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Perform phylogenetic PCA</span>
phyl_pca <- <span class="code_function" >phyl.pca</span>(tree, Y=gpa_mat, method=<span class="code_quote" >'BM'</span>)
</pre></div>

<p>
From the <span class='in_line_code'>phyl.pca</span> results you can get the pPCA scores, eigenvectors, and percent variance 
explained by each axis, just as for a non-phylogenetic PCA.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Get PC scores</span>
ppca_scores <- phyl_pca$S

<span class="code_comment" ># Get PC eigenvectors</span>
ppca_vectors <- phyl_pca$Evec

<span class="code_comment" ># Get percent variance explained along each axis</span>
diag_values <- phyl_pca$Eval[<span class="code_function" >diag</span>(<span class="code_function" >nrow</span>(phyl_pca$Eval)) == TRUE]
ppca_per_var <- (diag_values / <span class="code_function" >sum</span>(diag_values, na.rm=TRUE))*<span class="code_numeric" >100</span>
</pre></div>

<p>
Lastly get a vector of the phylogenetic means using the covariance and variance-covariance 
matrices. This will be input into <span class='in_line_code'>btShapes</span>.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Find covariance matrix due to phylogenetic relatedness</span>
cov_mat <- <span class="code_function" >vcv.phylo</span>(tree)[<span class="code_function" >rownames</span>(gpa_mat), <span class="code_function" >rownames</span>(gpa_mat)]

<span class="code_comment" ># Compute the phylogenetic trait variance-covariance matrix to get the phylogenetic means</span>
phy_means <- <span class="code_function" >phyl.vcv</span>(gpa_mat, cov_mat, <span class="code_numeric" >1</span>)$alpha
</pre></div>

<p>
Set the two PC axes that you want to plot, just as in the previous section.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Set PCs to plot</span>
pcs <- <span class="code_numeric" >1</span>:<span class="code_numeric" >2</span>
</pre></div>

<p>
Open a PDF graphics device where the plot will be produced and create a plot box with axes 
and axis labels. Use the <span class='in_line_code'>n</span> value for the <span class='in_line_code'>type</span> parameter to draw the plot without 
plotting the scores. That way you can add the scores later on top of the backtransform shapes.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Open PDF graphics device</span>
<span class="code_function" >pdf</span>(<span class="code_quote" >'Backtransform pPCA.pdf'</span>, width=<span class="code_numeric" >9</span>, height=<span class="code_numeric" >6.5</span>)

<span class="code_comment" ># Create plot box with axes and axis labels</span>
<span class="code_function" >plot</span>(ppca_scores[, pcs], type=<span class="code_quote" >'n'</span>, main=<span class="code_quote" >'Phylogenetic backtransform morphospace'</span>,
	xlab=<span class="code_function" >paste0</span>(<span class="code_quote" >'pPC'</span>, pcs[<span class="code_numeric" >1</span>], <span class="code_quote" >' ('</span>, <span class="code_function" >round</span>(ppca_per_var[pcs[<span class="code_numeric" >1</span>]]), <span class="code_quote" >'%)'</span>),
	ylab=<span class="code_function" >paste0</span>(<span class="code_quote" >'pPC'</span>, pcs[<span class="code_numeric" >2</span>], <span class="code_quote" >' ('</span>, <span class="code_function" >round</span>(ppca_per_var[pcs[<span class="code_numeric" >2</span>]]), <span class="code_quote" >'%)'</span>))
</pre></div>

<p>
Plot the backtransform shapes into the current plot using the <span class='in_line_code'>btShapes</span> function. This 
function call uses the same input parameters as for the non-phylogenetic PCA with one additional 
parameter for the phylogenetic means, <span class='in_line_code'>phy.means</span>. Note that now you'll input the <i>pPCA</i> scores 
and vectors instead of the PCA scores and vectors.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Plot backtransform shapes</span>
<span class="code_function" >btShapes</span>(scores=ppca_scores, vectors=ppca_vectors, fcn=plot_beak_lateral, 
	pcs=pcs, n=<span class="code_function" >c</span>(<span class="code_numeric" >6</span>,<span class="code_numeric" >8</span>), m=<span class="code_function" >dim</span>(lm_array)[<span class="code_numeric" >2</span>], row.names=<span class="code_function" >dimnames</span>(lm_array)[[<span class="code_numeric" >1</span>]], 
	pc.margin=<span class="code_function" >c</span>(<span class="code_numeric" >0.06</span>,<span class="code_numeric" >0.05</span>), phy.means=phy_means, size=<span class="code_numeric" >0.5</span>, col=<span class="code_function" >gray</span>(<span class="code_numeric" >0.7</span>))
</pre></div>

<p>
Add the scores and labels onto the plot, if you'd like.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Plot points for each species</span>
<span class="code_function" >points</span>(ppca_scores[, pcs])

<span class="code_comment" ># Add text labels</span>
<span class="code_function" >text</span>(ppca_scores[, pcs], labels=<span class="code_function" >substr</span>(<span class="code_function" >rownames</span>(ppca_scores), <span class="code_numeric" >0</span>, <span class="code_numeric" >3</span>), cex=<span class="code_numeric" >0.8</span>, pos=<span class="code_numeric" >1</span>, offset=<span class="code_numeric" >0.3</span>)
</pre></div>

<p>
And close the connection to the PDF graphics device.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Close the PDF graphics device</span>
<span class="code_function" >dev.off</span>()
</pre></div>

<p>
This should produce the following plot:
</p>

<div class='p_image_middle_div' style="width:550px;">
	<div><a href='../../img/tutorials/morphometrics/backtransform_morphospace_ppca.jpg' target='_blank'>
		<img width='550px' src='../../img/tutorials/morphometrics/backtransform_morphospace_ppca.jpg' /></a></div>
	<div class='p_image_caption' style='text-align:center;' >
		Phylogenetic backtransform morphospace of 3D waterfowl beak shapes
	</div>
</div>

<p>
It's interesting to now compare the two plots side by side. Looking at the relative positions 
of the points, adding the phylogeny the PCA hasn't substantially changed the positions of the points relative 
to one another. Rather it has taken all of the points and rotated them about 30-45 degrees relative 
to the center of the plot. This has the effect of changing the axes of shape variation that 
are orthogonal (uncorrelated) with one another.
</p>

<p>
For this particular dataset there is an interesting phylogenetic pattern in that a more goose-like 
beak (more toward the left in the pPCA morphospace) has evolved convergently multiple times 
in waterfowl, probably at least six times (<a href='https://doi.org/10.1111/1365-2435.12890' target='_blank'>Olsen 2017</a>). 
Once phylogeny is taken into account the variation between a more duck-like beak (probably ancestral) and 
a more goose-like beak weighs more heavily on the PCA (because they represent more 'independent' transitions). 
The end effect is that the axis differentiating duck and goose beaks is more squarely aligned 
with the first PC axis in the pPCA (right) than in the PCA (left).
</p>

<div>
	<div class='p_image_middle_div' style="width:700px;position:float;">
		<div style="float:left" >
			<img width='350px' src='../../img/tutorials/morphometrics/backtransform_morphospace.jpg' />
			<div class='p_image_caption' style='text-align:center;' >Non-phylogenetic PCA</div>
		</div>
		<div style="float:right">
			<img width='350px' src='../../img/tutorials/morphometrics/backtransform_morphospace_ppca.jpg' />
			<div class='p_image_caption' style='text-align:center;' >Phylogenetic PCA</div>
		</div>
	</div>
</div>

<a name="background_morphospace_2d" ></a>

<h2>Creating a backtransform morphospace using 2D shape data</h2>

<p>
Creating a backtransform morphospace using 2D shape data only requires changes 
to the custom function that draws each backtransform shape (the <span class='in_line_code'>fcn</span> input to 
btShapes). You can easily convert the example dataset from this tutorial into 
a 2D dataset by taking only the first two dimensions. Since the beaks are initially 
aligned from front to back in the xy-plane this removes 
any data about the width of the beak. After converting the shape data into an 
array 
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Convert list of shape data into array</span>
lm_array <- <span class="code_function" >array</span>(<span class="code_function" >unlist</span>(read_files), dim=<span class="code_function" >c</span>(<span class="code_function" >dim</span>(read_files[[<span class="code_numeric" >1</span>]]), 
	<span class="code_function" >length</span>(read_files)), dimnames=<span class="code_function" >list</span>(<span class="code_function" >rownames</span>(read_files[[<span class="code_numeric" >1</span>]]), 
	NULL, <span class="code_function" >gsub</span>(<span class="code_quote" >'[.]txt$'</span>, <span class="code_quote" >''</span>, <span class="code_function" >list.files</span>(<span class="code_quote" >'Beak shape by species'</span>))))
</pre></div>

<p>
remove the last dimension
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Take only the first two dimensions</span>
lm_array <- lm_array[, <span class="code_numeric" >1</span>:<span class="code_numeric" >2</span>, ]
</pre></div>

<p>
Then you can run all of the subsequent code as before (you'll need to reduce the <span class='in_line_code'>size</span> parameter in btShapes from 0.5 to 0.35).
Note that the <span class='in_line_code'>m</span> input parameter to btShapes should now be 2 versus 3. Using 
<span class='in_line_code'>dim(lm_array)[2]</span> will automatically account for this by reading the number of 
coordinate dimensions in lm_array.
</p>

<div class='p_image_middle_div' style="width:550px;">
	<div><a href='../../img/tutorials/morphometrics/backtransform_morphospace_2d.jpg' target='_blank'>
		<img width='550px' src='../../img/tutorials/morphometrics/backtransform_morphospace_2d.jpg' /></a></div>
	<div class='p_image_caption' style='text-align:center;' >
		Backtransform morphospace of 2D waterfowl beak shapes
	</div>
</div>

<p>
The plot above shows the result of using 2D shape data. The backtransform shapes looks nearly 
identical because we've used the same two dimensions that we plotted last time (again, 
ignoring beak width and any shapes in that dimension). So we haven't lost any 
information for the 2D plane in which we are visualizing the shapes. This wouldn't 
be the case if we had visualized the beaks from top down.
</p>

<p>
The percent variance explained 
has changed a bit for each axis. And note that the positions of the various species 
in the plot has flipped along the x-axis ('biz' for Biziura lobata, the musk duck, has moved from 
the right to the left side). This doesn't have any special significance; 
the orientation or order of the PC scores along each axis 
in PCA is arbitrary and sometimes the axes get flipped when slight changes are 
made to the data.
</p>

<h2>Citing this tutorial</h2>

<p>
If you are using this code for a publication, please cite the following paper:
</p>

<div style='margin:0 0 20px 20px;padding-left:10px;'>
Olsen AM (2017). Feeding ecology is the primary driver of beak shape diversification in waterfowl. 
<em>Functional Ecology</em>. 31(10):1985-1995. DOI: <a href='https://doi.org/10.1111/1365-2435.12890' target='_blank'>10.1111/1365-2435.12890</a>. Get <a href='http://onlinelibrary.wiley.com/doi/10.1111/1365-2435.12890/full' target='_blank' >via Wiley</a>, 
<a href='https://www.researchgate.net/publication/317316889_Feeding_ecology_is_the_primary_driver_of_beak_shape_diversification_in_waterfowl' target='_blank' > via ResearchGate</a>.
</div>

<p>
Be sure to cite the <a href="https://CRAN.R-project.org/package=geomorph" target="_blank" >geomorph</a> package and 
the <a href="https://cran.r-project.org/doc/FAQ/R-FAQ.html#Citing-R" target="_blank" >R Core Team</a>. 
And if you created a pPCA backtransform morphospace then you should also cite 
<a href="https://CRAN.R-project.org/package=ape" target="_blank" >ape</a>,  
<a href="https://CRAN.R-project.org/package=phytools" target="_blank" >phytools</a>, and 
<a href='https://doi.org/10.1111/j.1558-5646.2009.00804.x' target='_blank' >Revell 2009</a>. 
You can generate the most up-to-date citations for R packages using the 'citation' function.
</p>

<div class="code" ><pre class="code">
<span class="code_comment" ># Get current citation for an R package, geomorph for example</span>
<span class="code_function" >citation</span>(<span class="code_quote" >'geomorph'</span>)
</pre></div>

<p>
If you find an error or have a question, please <a href='../../contact.html' target='_blank'> contact me</a>!
</p>
</div>

		</div>

		<div id='footer' class='footer' >
			<div style='float:left;' >Â© 2019 Aaron Olsen. All rights reserved.<br><br></div>
			<div style='float:right;' >Design and photographs by Aaron Olsen.</div>
			<div style='float:left;background-color:;' >The material on this site is based upon work supported by the National Science Foundation (DGE-1144082, DGE-0903637, DBI-1612230). Any opinions, findings, and conclusions or recommendations expressed in this material are those of the author(s) and do not necessarily reflect the views of the National Science Foundation.</div>
		</div>

	</div>

</body>
<script src="../../js/sharedfunctions2.js" type="text/javascript" ></script>